gms/gossiper.hh

```diff
@@ -84,6 +85,45 @@ class i_endpoint_state_change_subscriber;
 
 class feature_service;
 
+class gossip_query_token_status_response {
```

Comment:
Please add a comment above this class which describes shortly what it is and what it's used for.
E.g. it is a "summary of the ring", used for "deciding whether all nodes have the same view of the ring"

Reply:
Also, I think that the name of the class introduces a leak between different abstraction levels.
The class shouldn't care that it's used inside gossip and that it's a "response" to something.

It should only care what information/data structure it represents. So IMO it should be called something like "token_status", or more generally, "ring_status", or even better, "ring_status_summary" because it doesn't have all the info but it's just a hash.

You can however, and should describe in comments what the class is used for, because it's helpful for understanding the entire thing in context.

Reply:
Yes, I liked the name ring_status_summary. It is true it is an abstraction not just gossip. It is a combination of both gossip status and the tokens status. 

I am putting this in gossip because gossiper::wait_for_gossip_to_settle() and gossiper::wait_for_range_setup() which consumes gossip_query_token_status_response are all in the gossip.

In my view, it is an overkill to introduce a new service just for this.

Reply:
> In my view, it is an overkill to introduce a new service just for this.

It wouldn't have to be a new service, it would be ok for a standalone function. Do we actually need to access some gossip internals or can we somehow separate the two abstractions? But it's your decision in the end.

> I am putting this in gossip because gossiper::wait_for_gossip_to_settle() and gossiper::wait_for_range_setup() which consumes gossip_query_token_status_response are all in the gossip.

I wonder if it's all right to use `wait_for_gossip_with_query()` inside `wait_for_gossip_to_settle()`.
Wasn't `wait_for_gossip_to_settle()` called even before we synchronized token information? I.e. the joining node doesn't inform other nodes of its chosen tokens yet, but it calls `wait_for_gossip_to_settle()` to make sure we've reached schema feature agreement.

Reply:
gossiper::wait_for_gossip_with_query access endpoint_state_map and _seed in side gossiper but we can access them outside gossiper too.  Btw, the handler get_gossip_query_token_status_response is already a static function. I will think about if we make gossiper::wait_for_gossip_with_query a free function too. Thanks.

Reply:
The call you mentioned is:
```
 443     // Wait for gossip to settle so that the fetures will be enabled
 444     if (do_bind) {
 445         gms::get_local_gossiper().wait_for_gossip_to_settle().get();
 446     }
```
This is used to make sure gossip is settled so that the features can be are enabled correctly. 
The responses inside wait_for_gossip_with_query should be the same for a given state of the cluster (e.g., after the boot node adding its status to bootstrap, to normal, etc. ). 
There is no specific point to "sync token information".